File: python/src/chaos.py
"""Chaos experiments as CI-runnable SLO assertions.

Inject deterministic faults (added latency, periodic errors) into a stream of
observed call latencies, then assert that Service Level Objectives still hold. If
they don't, the experiment fails -- resilience testing as a pass/fail gate, not a
one-off manual drill.

Deterministic by design (fault schedule, not RNG) so results are reproducible and
identical across Python, C#, and Java.
"""

from __future__ import annotations

import math
from dataclasses import dataclass


@dataclass(frozen=True)
class FaultPolicy:
    added_latency_ms: float = 0.0
    fail_every_n: int = 0  # 0 = never; else fail calls where (index % n == 0), 1-indexed


@dataclass(frozen=True)
class ExperimentResult:
    total_calls: int
    failures: int
    success_rate: float
    p95_latency_ms: float


@dataclass(frozen=True)
class Slo:
    min_success_rate: float
    max_p95_latency_ms: float


@dataclass(frozen=True)
class SloResult:
    passed: bool
    success_rate: float
    p95_latency_ms: float
    breaches: list[str]


def percentile(values: list[float], p: float) -> float:
    """Nearest-rank percentile (p in [0,1])."""
    if not values:
        return 0.0
    ordered = sorted(values)
    rank = math.ceil(p * len(ordered))
    idx = min(max(rank - 1, 0), len(ordered) - 1)
    return ordered[idx]


def run_experiment(base_latencies_ms: list[float], policy: FaultPolicy) -> ExperimentResult:
    latencies: list[float] = []
    failures = 0
    for i, base in enumerate(base_latencies_ms, start=1):
        is_failure = policy.fail_every_n > 0 and i % policy.fail_every_n == 0
        if is_failure:
            failures += 1
            continue  # failed calls don't contribute a latency sample
        latencies.append(base + policy.added_latency_ms)

    total = len(base_latencies_ms)
    successes = total - failures
    return ExperimentResult(
        total_calls=total,
        failures=failures,
        success_rate=round(successes / total, 6) if total else 0.0,
        p95_latency_ms=round(percentile(latencies, 0.95), 6),
    )


def evaluate_slo(result: ExperimentResult, slo: Slo) -> SloResult:
    breaches = []
    if result.success_rate < slo.min_success_rate:
        breaches.append(
            f"success_rate {result.success_rate:.4f} < {slo.min_success_rate:.4f}"
        )
    if result.p95_latency_ms > slo.max_p95_latency_ms:
        breaches.append(
            f"p95 {result.p95_latency_ms:.2f}ms > {slo.max_p95_latency_ms:.2f}ms"
        )
    return SloResult(
        passed=not breaches,
        success_rate=result.success_rate,
        p95_latency_ms=result.p95_latency_ms,
        breaches=breaches,
    )

File: csharp/Chaos.cs
namespace ChaosMeshLite;

public readonly record struct FaultPolicy(double AddedLatencyMs = 0.0, int FailEveryN = 0);

public readonly record struct ExperimentResult(
    int TotalCalls, int Failures, double SuccessRate, double P95LatencyMs);

public readonly record struct Slo(double MinSuccessRate, double MaxP95LatencyMs);

public sealed record SloResult(bool Passed, double SuccessRate, double P95LatencyMs, List<string> Breaches);

/// <summary>Chaos experiments as CI-runnable SLO assertions (deterministic fault schedule).</summary>
public static class Chaos
{
    public static double Percentile(IReadOnlyList<double> values, double p)
    {
        if (values.Count == 0)
        {
            return 0.0;
        }

        var ordered = values.OrderBy(v => v).ToList();
        var rank = (int)Math.Ceiling(p * ordered.Count);
        var idx = Math.Min(Math.Max(rank - 1, 0), ordered.Count - 1);
        return ordered[idx];
    }

    public static ExperimentResult RunExperiment(IReadOnlyList<double> baseLatenciesMs, FaultPolicy policy)
    {
        var latencies = new List<double>();
        var failures = 0;
        for (var i = 0; i < baseLatenciesMs.Count; i++)
        {
            var oneIndexed = i + 1;
            var isFailure = policy.FailEveryN > 0 && oneIndexed % policy.FailEveryN == 0;
            if (isFailure)
            {
                failures++;
                continue;
            }

            latencies.Add(baseLatenciesMs[i] + policy.AddedLatencyMs);
        }

        var total = baseLatenciesMs.Count;
        var successes = total - failures;
        return new ExperimentResult(
            total,
            failures,
            total == 0 ? 0.0 : Math.Round((double)successes / total, 6),
            Math.Round(Percentile(latencies, 0.95), 6));
    }

    public static SloResult EvaluateSlo(ExperimentResult result, Slo slo)
    {
        var breaches = new List<string>();
        if (result.SuccessRate < slo.MinSuccessRate)
        {
            breaches.Add($"success_rate {result.SuccessRate:F4} < {slo.MinSuccessRate:F4}");
        }

        if (result.P95LatencyMs > slo.MaxP95LatencyMs)
        {
            breaches.Add($"p95 {result.P95LatencyMs:F2}ms > {slo.MaxP95LatencyMs:F2}ms");
        }

        return new SloResult(breaches.Count == 0, result.SuccessRate, result.P95LatencyMs, breaches);
    }
}
